modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/upgrade/v3_0_2/RedirectEntrySourceURLUpgradeProcess.java

```diff
@@ -0,0 +1,125 @@
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Attila Bakay
+ */
+public class RedirectEntrySourceURLUpgradeProcess extends UpgradeProcess {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement preparedStatement1 = connection.prepareStatement(
+				"select redirectEntryId, groupId, sourceURL from " +
+					"RedirectEntry ORDER BY redirectentryid ASC");
+			PreparedStatement preparedStatement2 =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update RedirectEntry set sourceURL = ? where " +
+						"redirectEntryId = ?")) {
+
+			ResultSet resultSet = preparedStatement1.executeQuery();
+
+			while (resultSet.next()) {
+				long redirectEntryId = resultSet.getLong(1);
+				long groupId = resultSet.getLong(2);
+				String sourceURL = resultSet.getString(3);
+
+				Set<String> groupSpecificSourceURLs;
+				Map<String, Long> groupSpecificUpgrades;
+
+				if (!_redirectEntries.containsKey(groupId)) {
+					groupSpecificSourceURLs = new HashSet<>();
+					groupSpecificUpgrades = new LinkedHashMap<>();
+				}
+				else {
+					groupSpecificSourceURLs = _redirectEntries.get(groupId);
+					groupSpecificUpgrades = _upgradeEntries.get(groupId);
+				}
+
+				String lowerCaseSourceURL = StringUtil.toLowerCase(sourceURL);
+
+				if (!groupSpecificSourceURLs.contains(lowerCaseSourceURL)) {
+					groupSpecificSourceURLs.add(lowerCaseSourceURL);
+
+					if (!sourceURL.equals(lowerCaseSourceURL)) {
+						groupSpecificUpgrades.put(
+							lowerCaseSourceURL, redirectEntryId);
+					}
+
+					_upgradeEntries.put(groupId, groupSpecificUpgrades);
+					_redirectEntries.put(groupId, groupSpecificSourceURLs);
+				}
+				else {
+					if (sourceURL.equals(lowerCaseSourceURL)) {
+						groupSpecificUpgrades.remove(lowerCaseSourceURL);
+					}
+					else {
+						if (_log.isWarnEnabled()) {
+							_log.warn(
+								StringBundler.concat(
+									"Can not modify '", sourceURL, "' to '",
+									lowerCaseSourceURL,
+									"' because it is already in use in group: ",
+									groupId, " at id: ", redirectEntryId));
+						}
+					}
+				}
+			}
+
+			for (Map<String, Long> groupSpecificUpgrades :
+					_upgradeEntries.values()) {
+
+				for (Map.Entry<String, Long> upgradeRedirectEntry :
+						groupSpecificUpgrades.entrySet()) {
+
+					String lowerCaseSourceURL = upgradeRedirectEntry.getKey();
+					long redirectEntryId = upgradeRedirectEntry.getValue();
+
+					preparedStatement2.setString(1, lowerCaseSourceURL);
+					preparedStatement2.setLong(2, redirectEntryId);
+
+					preparedStatement2.addBatch();
+				}
+			}
+
+			preparedStatement2.executeBatch();
+		}
+	}
+
+	private static final Log _log = LogFactoryUtil.getLog(
+		RedirectEntrySourceURLUpgradeProcess.class);
+
+	private final Map<Long, Set<String>> _redirectEntries =
+		new LinkedHashMap<>();
+	private final Map<Long, Map<String, Long>> _upgradeEntries =
+		new LinkedHashMap<>();
```

Comment:
Why not save the redirect entries themselves instead of a Map<String, Long>? If you're doing it to avoid lowcasing the url twice, that's not a problem, and the code would be clearer. If you decide to do this, remember to rename the variable to `redirectEntries` or something that directly hints the kind of thing it contains.

Reply:
Hi @adolfopa,
Do you mean I should use a RedirectEntires collection instead of only using the entryId and sourceURL for maps? I mapped them this way because it is faster if I use the sourceURL as a key, and I can use that because at this point I already handle all the groups in different collections.

Reply:
I see. It makes sense, so it forget about it.

In any case, you'll need to find a better variable name.